File: Implementation/Test/Harness/C#/Harness.cs
using Console = System.Console;
using IO = System.IO;
using Cks.Data;
using Cks.Text.Writer;
using Cks.Text.Reader;
using Cks.Binary.Writer;
using Cks.Binary.Reader;
using Cks.Text.Reader.Model;

public class Harness
{
	private sealed class InputSpec
	{
		public readonly FormatT Format;
		public readonly string File;

		public enum FormatT { Text, Binary, }

		public InputSpec(FormatT Format, string File)
		{
			this.Format = Format;
			this.File = File;
		}
	}

	private sealed class OutputSpec
	{
		public readonly FormatT Format;
		public readonly string File;

		public enum FormatT { Text, TextIndented, Binary, }

		public OutputSpec(FormatT Format, string File)
		{
			this.Format = Format;
			this.File = File;
		}
	}

	public static void Run<T>(
		string[] Args,
		BinaryReader<T> BinaryReader,
		BinaryWriter<T> BinaryWriter,
		TextReader<T> TextReader,
		TextWriter<T> TextWriter)
	{
		if (Args.Length != 0) {
			Console.Error.WriteLine("! Expecting zero arguments.");
			System.Environment.Exit(1);
		}

		while (true) {
			string Command = Console.ReadLine();
			if (Command == null) break;

			try {
				if (Command.Equals("convert")) {
					string InputString = Console.ReadLine();
					string OutputString = Console.ReadLine();

					InputSpec Input = ParseInputSpec(InputString, "input-spec");
					OutputSpec Output = ParseOutputSpec(OutputString, "output-spec");

					T Value = ReadValue(Console.Error, Input, BinaryReader, TextReader);
					WriteValue(Console.Error, Output, BinaryWriter, TextWriter, Value);
				}
				else if (Command.Equals("compare")) {
					string InputString1 = Console.ReadLine();
					string InputString2 = Console.ReadLine();

					InputSpec Input1 = ParseInputSpec(InputString1, "first input-spec");
					InputSpec Input2 = ParseInputSpec(InputString2, "second input-spec");


					T Value1 = ReadValue(Console.Error, Input1, BinaryReader, TextReader);
					T Value2 = ReadValue(Console.Error, Input2, BinaryReader, TextReader);

					Console.Out.WriteLine("> " + (Value1.Equals(Value2) ? "equal" : "not equal"));
				}
				else {
					Console.Error.WriteLine("! Unknown command: \"" + Command + "\"");
					System.Environment.Exit(1); break;
				}
			}
			catch (ExitException) {
				// Ignore it.
			}
			finally {
				Console.Out.WriteLine(">.");
				Console.Error.WriteLine("!.");
			}
		}
	}

	private static InputSpec ParseInputSpec(string Spec, string Name)
	{
		int ColonPos = Spec.IndexOf(':');
		if (ColonPos < 0) {
			Console.Error.WriteLine("! Invalid " + Name + ": missing colon");
			throw Exit(1);
		}

		string Format = Spec.Substring(0, ColonPos);
		string File = Spec.Substring(ColonPos+1);

		InputSpec.FormatT f;

		if (Format.Equals("text")) {
			f = InputSpec.FormatT.Text;
		}
		else if (Format.Equals("binary")) {
			f = InputSpec.FormatT.Binary;
		}
		else {
			Console.Error.WriteLine("! Invalid " + Name + ": \"" + Format + "\" is not a valid input format");
			throw Exit(1);
		}

		return new InputSpec(f, File);
	}

	private static OutputSpec ParseOutputSpec(string Spec, string Name)
	{
		int ColonPos = Spec.IndexOf(':');
		if (ColonPos < 0) {
			Console.Error.WriteLine("! Invalid " + Name + ": missing colon");
			throw Exit(1);
		}

		string Format = Spec.Substring(0, ColonPos);
		string File = Spec.Substring(ColonPos+1);

		OutputSpec.FormatT f;

		if (Format.Equals("text")) {
			f = OutputSpec.FormatT.Text;
		}
		else if (Format.Equals("texti")) {
			f = OutputSpec.FormatT.TextIndented;
		}
		else if (Format.Equals("binary")) {
			f = OutputSpec.FormatT.Binary;
		}
		else {
			Console.Error.WriteLine("! Invalid " + Name + ": \"" + Format + "\" is not a valid output format");
			throw Exit(1);
		}

		return new OutputSpec(f, File);
	}

	private static T ReadValue<T>(IO.TextWriter Err, InputSpec Input, BinaryReader<T> BinaryReader, TextReader<T> TextReader)
	{
		T Value;
		if (Input.Format == InputSpec.FormatT.Text) {
			Value = ReadText(Console.Error, Input.File, TextReader);
		}
		else if (Input.Format == InputSpec.FormatT.Binary) {
			Value = ReadBinary(Console.Error, Input.File, BinaryReader);
		}
		else {
			throw new System.ApplicationException("bad type: " + Input.Format);
		}
		return Value;
	}

	private static void WriteValue<T>(IO.TextWriter Err, OutputSpec Output, BinaryWriter<T> BinaryWriter, TextWriter<T> TextWriter, T Value)
	{
		if (Output.Format == OutputSpec.FormatT.Text) {
			WriteText(Console.Error, Output.File, TextWriter, Value);
		}
		else if (Output.Format == OutputSpec.FormatT.TextIndented) {
			WriteTextIndented(Console.Error, Output.File, TextWriter, Value);
		}
		else if (Output.Format == OutputSpec.FormatT.Binary) {
			WriteBinary(Console.Error, Output.File, BinaryWriter, Value);
		}
		else {
			throw new System.ApplicationException("bad type: " + Output.Format);
		}
	}

	private static T ReadBinary<T>(IO.TextWriter Err, string File, BinaryReader<T> BinaryReader)
	{
		try {
			IO.FileStream fs = IO.File.OpenRead(File);
			using (IO.BinaryReader In = new IO.BinaryReader(fs)) {
				T v = BinaryReader.Read(In);
				if (In.Read() != -1) {
					int ExtraBytes = 1;
					while (In.Read() != -1) ExtraBytes++;
					Err.WriteLine("! Error: extra bytes after payload: " + ExtraBytes + ".");
					throw Exit(2);
				}
				return v;
			}
		}
		catch (IO.IOException e) {
			Err.WriteLine("! Error reading from file \"" + File + "\": " + e.Message);
			throw Exit(2);
		}
		catch (BinaryFormatException e) {
			Err.WriteLine("! Binary format error: " + e.Message);
			throw Exit(2);
		}
	}

	private static T ReadText<T>(IO.TextWriter Err, string File, TextReader<T> TextReader)
	{
		try {
			using (IO.TextReader In = new IO.StreamReader(File)) {
				return TextReader.Read(In);
			}
		}
		catch (ProblemException e) {
			Err.WriteLine("! Marshal error: " + e.Problem);
			throw Exit(2);
		}
	}

	private static void WriteText<T>(IO.TextWriter Err, string File, TextWriter<T> TextWriter, T Value)
	{
		try {
			using (IO.TextWriter Out = new IO.StreamWriter(File)) {
				TextWriter.Write(Out, Value);
			}
		}
		catch (IO.IOException e) {
			Err.WriteLine("! Error writing to file \"" + File + "\": " + e.Message);
			throw Exit(2);
		}
	}

	private static void WriteTextIndented<T>(IO.TextWriter Err, string File, TextWriter<T> TextWriter, T Value)
	{
		try {
			using (IO.TextWriter Out = new IO.StreamWriter(File)) {
				TextWriter.WriteIndented(new IndentPrinter(Out), Value);
			}
		}
		catch (IO.IOException e) {
			Err.WriteLine("! Error writing to file \"" + File + "\": " + e.Message);
			throw Exit(2);
		}
	}

	private static void WriteBinary<T>(IO.TextWriter Err, string File, BinaryWriter<T> BinaryWriter, T Value)
	{
		try {
			IO.FileStream fs = IO.File.Open(File, IO.FileMode.Create, IO.FileAccess.Write);
			using (IO.BinaryWriter Out = new IO.BinaryWriter(fs)) {
				BinaryWriter.Write(Out, Value);
			}
		}
		catch (IO.IOException e) {
			Err.WriteLine("! Error writing to file \"" + File + "\": " + e.Message);
			throw Exit(2);
		}
	}

	private static System.ApplicationException Exit(int Code)
	{
		return new ExitException(Code);
	}

	private class ExitException : System.ApplicationException
	{
		public readonly int Code;
		public ExitException(int Code)
		{
			this.Code = Code;
		}
	}

}

File: Implementation/Test/Harness/Haskell/Harness.hs
module Harness (run) where

import Prelude hiding (catch)
import qualified Data.Cks.TreeDoc as TD
import qualified Data.Cks.TextParser as TextParser
import qualified Data.Cks.TextReader as TextReader
import qualified Data.Cks.TextModel as TextModel
import qualified Data.Cks.BinaryReader as BinaryReader
import qualified Data.Cks.BinaryWriter as BinaryWriter
import Data.Cks.TextReader (TextMarshaller)
import Data.Cks.TextWriter (TextWriter, TextIndentedWriter)
import Data.Cks.BinaryReader (BinaryReader)
import Data.Cks.BinaryWriter (BinaryWriter)
import Data.Cks.MaybeE

import qualified Data.ByteString as B

import Control.Exception (catch, finally)
import System.Exit (exitWith, ExitCode(ExitFailure))
import System.IO (hPutStr, hPutStrLn, hGetContents, stdin, stderr, stdout,
	withFile, withBinaryFile, IOMode(WriteMode, ReadMode), hIsClosed, hFlush)
import System.Environment (getArgs)
import Text.ParserCombinators.Parsec (runParser)
import Data.List (elemIndex)

run :: Ord a =>
	   BinaryReader a
	-> BinaryWriter a
	-> TextMarshaller a
	-> TextWriter a
	-> TextIndentedWriter a
	-> IO ()
run br bw tr tw tiw = do
	closed <- hIsClosed stdin
	if closed
		then return ()
		else do
			(run' br bw tr tw tiw `catch` ignore) `finally` (do
				hPutStrLn stdout ">."
				hFlush stdout
				hPutStrLn stderr "!."
				hFlush stderr)
			run br bw tr tw tiw
	where
		ignore (e :: ExitCode) = return ()

run' :: Ord a =>
	   BinaryReader a
	-> BinaryWriter a
	-> TextMarshaller a
	-> TextWriter a
	-> TextIndentedWriter a
	-> IO ()
run' binaryReader binaryWriter textReader textWriter textIndentedWriter = do
	command <- getLine
	case command of
		"convert" -> do
			inputString <- getLine
			outputString <- getLine
			inputSpec <- parseOrDie "input-spec" (parseInputSpec inputString)
			outputSpec <- parseOrDie "output-spec" (parseOutputSpec outputString)
			value <- readValue inputSpec
			writeValue outputSpec value
		"compare" -> do
			inputString1 <- getLine
			inputString2 <- getLine
			inputSpec1 <- parseOrDie "input-spec-1" (parseInputSpec inputString1)
			inputSpec2 <- parseOrDie "input-spec-2" (parseInputSpec inputString2)
			value1 <- readValue inputSpec1
			value2 <- readValue inputSpec2
			hPutStrLn stdout $ "> " ++ (if value1 == value2 then "equal" else "not equal")
		_ -> do
			hPutStrLn stderr $ "! Invalid command: \"" ++ command ++ "\""
			exitWith $ ExitFailure 1

	where
		readValue (format, file) = case format of
			IF_Text -> withFile file ReadMode (readText file)
			IF_Binary -> withBinaryFile file ReadMode (readBinary file)
		writeValue (format, file) v = case format of
			OF_Text -> withFile file WriteMode (writeText v)
			OF_TextI -> withFile file WriteMode (writeTextIndented v)
			OF_Binary -> withBinaryFile file WriteMode (writeBinary v)
		writeBinary v f = case (binaryWriter v) of
			Error err -> do
				hPutStrLn stderr $ "! Error writing value in binary format: " ++ err
				exitWith $ ExitFailure 1
			Result bytes -> do B.hPutStr f bytes
		writeText v f = case (textWriter v) of
			Error err -> do
				hPutStrLn stderr $ "! Error writing value in text format: " ++ err
				exitWith $ ExitFailure 1
			Result chars -> do hPutStr f chars
		writeTextIndented v f = case (textIndentedWriter v) of
			Error err -> do
				hPutStrLn stderr $ "! Error writing value in text format: " ++ err
				exitWith $ ExitFailure 1
			Result doc -> do hPutStr f (TD.print "\t" "" doc)
		readBinary fn f = do
			bytes <- B.hGetContents f
			case (BinaryReader.run binaryReader bytes) of
				Error err -> do
					hPutStrLn stderr $ "! " ++ fn ++ ": " ++ err
					exitWith $ ExitFailure 1
				Result v -> return v
		readText fn f = do
			chars <- hGetContents f
			case TextReader.readText textReader chars of
				Error err -> do
					hPutStrLn stderr $ "! " ++ fn ++ ": " ++ show err
					exitWith $ ExitFailure 1
				Result v -> return v

type ErrorMessage = String

parseOrDie :: String -> MaybeE ErrorMessage a -> IO a
parseOrDie name result = case result of
	Error err -> do
		hPutStrLn stderr $ "! Invalid " ++ name ++ ": " ++ err
		exitWith $ ExitFailure 1
	Result v -> return v

parseInputSpec :: String -> MaybeE ErrorMessage InputSpec
parseInputSpec s =
	case elemIndex ':' s of
		Just colonPos ->
			let
				format = take colonPos s
				file = drop (colonPos+1) s
			in
				case format of
					"text" -> Result (IF_Text, file)
					"binary" -> Result (IF_Binary, file)
					_ -> Error $ "invalid input format: \"" ++ format ++ "\""
		Nothing -> Error "missing colon"

parseOutputSpec :: String -> MaybeE ErrorMessage OutputSpec
parseOutputSpec s =
	case elemIndex ':' s of
		Just colonPos ->
			let
				format = take colonPos s
				file = drop (colonPos+1) s
			in
				case format of
					"text" -> Result (OF_Text, file)
					"texti" -> Result (OF_TextI, file)
					"binary" -> Result (OF_Binary, file)
					_ -> Error $ "invalid output format: \"" ++ format ++ "\""
		Nothing -> Error "missing colon"

type InputSpec = (InputFormat, FilePath)
type OutputSpec = (OutputFormat, FilePath)

data InputFormat = IF_Text | IF_Binary
data OutputFormat = OF_Text | OF_TextI | OF_Binary
